### PR TITLE
Add runtime API and optimize collation generation/verification

### PR DESCRIPTION
**Pull request by thiolliere**
Fix https://github.com/paritytech/polkadot/issues/2879

* incease ParachainHost runtime API version with the changes:
  * a new method: `validation_code_hash`
  * the method `validation_code` now returns `ValidationCodeAndHash` instead of `ValidationCode`.
* the client support both Patachainhost version 1 and 2. With version 2 being more optimize:
  * collation generation only query the validation code hash from relay chain runtime (so no computation of the hash anymore)
  * collation validation gets the validation code hash and the validation code from the relay chain runtime (so no computation of the hash anymore)

BREAKING CHANGE:

* the method validation_code signature for ParachainHost API has changed.

TODO:

* [x] test again in polkadot-launch
* [x] ~maybe split runtime changes in one PR and client usage of new runtime API in another PR https://github.com/paritytech/polkadot/pull/3380~ used new version instead.

### CODE REVIEW

File: primitives/src/v1.rs [hunk 1]
```diff
@@ -885,10 +885,22 @@ sp_api::decl_runtime_apis! {
...
 
+		/// Fetch the hash of the validation code used by a para, making the given
+		/// `OccupiedCoreAssumption`.
+		///
+		/// Returns `None` if either the para is not registered or the assumption is `Freed`
+		/// and the para already occupies a core.
+		///
+		/// This is similar to [`Self::validation_code`] except it fetch the hash of the code.
+		#[skip_initialize_block]
+		fn validation_code_hash(para_id: Id, assumption: OccupiedCoreAssumption) -> Option<Hash>;
```
[the remark this answers is not in the document]
Comment (ordian) [reply primitives/src/v1.rs:902]: AFAIK, it's not a breaking change.

**Review by ordian — COMMENTED** [review]
I think it would make sense to cache this call as we do it for other calls in
`node/core/runtime-api/src/cache.rs`

Also add a function to `node/subsystem-util/src/lib.rs` in `specialize_requests!`.

File: primitives/src/v1.rs [hunk 2]
```diff
@@ -885,10 +885,22 @@ sp_api::decl_runtime_apis! {
...
 		fn validation_code(para_id: Id, assumption: OccupiedCoreAssumption)
 			-> Option<ValidationCode>;
 
+		/// Fetch the hash of the validation code used by a para, making the given
+		/// `OccupiedCoreAssumption`.
+		///
+		/// Returns `None` if either the para is not registered or the assumption is `Freed`
+		/// and the para already occupies a core.
+		///
+		/// This is similar to [`Self::validation_code`] except it fetch the hash of the code.
```
Comment (ordian) [inline primitives/src/v1.rs:900]:
```suggestion
		/// This is similar to [`Self::validation_code`] except it fetches the hash of the code.
```

File: primitives/src/v1.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bkchr) [reply primitives/src/v1.rs:902]: Not a breaking change and also not yet released anyway.

**Comment by thiolliere** [comment]
> I think it would make sense to cache this call as we do it for other calls in
> `node/core/runtime-api/src/cache.rs`
> 
> Also add a function to `node/subsystem-util/src/lib.rs` in `specialize_requests!`.

Ah I was thinking only implementing the runtime new API first and then the client but as it is not yet released then I do the client change in the client in this PR.

**Review by ordian — APPROVED** [review]
LGTM, the caching change can either go here or in the follow up PR which actually uses this API.

**Review by rphmeier — DISMISSED** [review]

**Comment by ordian** [comment]
@thiolliere could you please merge master?

**Comment by thiolliere** [comment]
master is merged but I needs to double check implementeters guide and current implement

**Review by ordian — APPROVED** [review]
Looks good.
As a side note, maybe it's worth renaming `fn hash` to `fn compute_hash` (when it actually computes it) to note the difference in computational complexity.

**Comment by thiolliere** [comment]
but is it still ok to change the ParachainHost runtime API like this?
* the fuction: `validation_code` has now a different result, maybe we should introduce `validation_code_and_hash` instead or create a new version. Even if there is no parachain maybe there is a way to make the client call this function and that could result in an error.
* the function: `validation_code_hash` is introduced, maybe we should first introduce it in the runtime and once all runtime have upgraded we can start using it in the client.

**Comment by ordian** [comment]
> but is it still ok to change the ParachainHost runtime API like this?
> 
>     * the fuction: `validation_code` has now a different result, maybe we should introduce `validation_code_and_hash` instead or create a new version. Even if there is no parachain maybe there is a way to make the client call this function and that could result in an error.
> 
>     * the function: `validation_code_hash` is introduced, maybe we should first introduce it in the runtime and once all runtime have upgraded we can start using it in the client.

Good points, I think you're right that we need to introduce `validation_code_and_hash` instead of changing `validation_code` (and remove the latter when the client changes are released) and separate the client changes as well.

**Comment by thiolliere** [comment]
I opened the PR https://github.com/paritytech/polkadot/pull/3380 for the runtime api changes, once in, I'll upgrade this PR, and when polkadot use new runtime we can merge

**Comment by bkchr** [comment]
> but is it still ok to change the ParachainHost runtime API like this?
> 
>     * the fuction: `validation_code` has now a different result, maybe we should introduce `validation_code_and_hash` instead or create a new version. Even if there is no parachain maybe there is a way to make the client call this function and that could result in an error.
> 
>     * the function: `validation_code_hash` is introduced, maybe we should first introduce it in the runtime and once all runtime have upgraded we can start using it in the client.

The runtime api supports versioning and I would strictly recommend to use it here.

First, bump the version of the runtime api. Then you can put the attribute `changed_in` over `validation_code`. The client site should be written in a way to make it workable with the old and new runtime api functions. (You can do this by checking for the version of the runtime api)

File: node/core/runtime-api/src/lib.rs [hunk 3]
```diff
@@ -342,8 +349,65 @@ where
 		Request::CheckValidationOutputs(para, commitments, sender) =>
 			query!(CheckValidationOutputs, check_validation_outputs(para, commitments), sender),
 		Request::SessionIndexForChild(sender) => query!(SessionIndexForChild, session_index_for_child(), sender),
-		Request::ValidationCode(para, assumption, sender) =>
-			query!(ValidationCode, validation_code(para, assumption), sender),
+		Request::ValidationCode(para, assumption, sender) => {
```
Comment (thiolliere) [inline node/core/runtime-api/src/lib.rs:352]: This the only calls to the runtime parachain host API `validate_code` that I could found. So there shouldn't be any other need to support both old and new version in the codebase as far as I can see.

Comment (ordian) [reply node/core/runtime-api/src/lib.rs:352]: Yes, all runtime API request go through runtime api subsystem.

**Review by ordian — APPROVED** [review]

File: primitives/src/v1/mod.rs [hunk 4]
```diff
@@ -930,14 +932,55 @@ sp_api::decl_runtime_apis! {
...
+		/// making the given `OccupiedCoreAssumption`.
+		///
+		/// Returns `None` if either the para is not registered or the assumption is `Freed`
+		/// and the para already occupies a core.
+		///
+		/// To fetch only the hash of the code use [`Self::validation_code_hash`].
+		///
+		/// # Warning
+		///
+		/// This function is only available since parachain host version 2.
```
[the remark this answers is not in the document]
Comment (thiolliere) [reply primitives/src/v1/mod.rs]: yes, I use a note instead, considering all interaction goes through the request stuff, I think it is fine.

**Comment by thiolliere** [comment]
@rphmeier maybe I can ask the review again considering the code and situation has changed a bit

**Comment by ordian** [comment]
@thiolliere this is still relevant, right? could you merge master?

**Comment by pepyakin** [comment]
Closing because stale

**Pull request closed without merging**